lib/IRGen/LoadableByAddress.cpp

```diff
@@ -0,0 +1,1538 @@
+
+private:
+  llvm::SetVector<SILFunction *> modFuncs;
+  llvm::SetVector<SILInstruction *> conversionInstrs;
+  llvm::SetVector<LoadInst *> loadInstrsOfFunc;
+  llvm::DenseSet<SILInstruction *> modApplies;
+};
+} // end anonymous namespace
+
+static void setInstrUsers(StructLoweringState &pass, AllocStackInst *allocInstr,
+                          SILValue instrOperand, StoreInst *store) {
+  SmallVector<Operand *, 8> uses(instrOperand->getUses());
+  for (Operand *userOp : uses) {
+    SILInstruction *user = userOp->getUser();
+    if (user == store) {
+      continue;
+    }
+    if (std::find(pass.applies.begin(), pass.applies.end(), user) !=
+        pass.applies.end()) {
+      // We can only modify pre-filtered (non-objc) apply sites
+      userOp->set(allocInstr);
+    } else if (StoreInst *storeUser = dyn_cast<StoreInst>(user)) {
+      // Optimization: replace with copy_addr to reduce code size
+      assert(std::find(pass.storeInstsToMod.begin(), pass.storeInstsToMod.end(),
+                       storeUser) == pass.storeInstsToMod.end() &&
+             "Did not expect this instr in storeInstsToMod");
+      SILBuilder copyBuilder(storeUser);
+      SILValue tgt = storeUser->getDest();
+      createOutlinedCopyCall(copyBuilder, allocInstr, tgt, pass);
+      storeUser->replaceAllUsesWith(tgt);
+      storeUser->getParent()->erase(storeUser);
+    }
+  }
+}
+
+static void allocateAndSetForInstrOperand(StructLoweringState &pass,
+                                          SILInstruction *instrOperand) {
+  assert(instrOperand->getType().isObject());
+  SILBuilder allocBuilder(pass.F->begin()->begin());
+  AllocStackInst *allocInstr = allocBuilder.createAllocStack(
+      instrOperand->getLoc(), instrOperand->getType());
+
+  auto II = instrOperand->getIterator();
+  ++II;
+  SILBuilder storeBuilder(II);
+  StoreInst *store = nullptr;
+  if (pass.F->hasQualifiedOwnership()) {
+    store = storeBuilder.createStore(instrOperand->getLoc(), instrOperand,
+                                     allocInstr, StoreOwnershipQualifier::Init);
+  } else {
+    store = storeBuilder.createStore(instrOperand->getLoc(), instrOperand,
+                                     allocInstr,
+                                     StoreOwnershipQualifier::Unqualified);
+  }
+
+  // Insert stack deallocations.
+  for (TermInst *termInst : pass.returnInsts) {
+    SILBuilder deallocBuilder(termInst);
+    deallocBuilder.createDeallocStack(allocInstr->getLoc(), allocInstr);
+  }
+
+  // Traverse all the uses of instrOperand - see if we can replace
+  setInstrUsers(pass, allocInstr, instrOperand, store);
+}
+
+static void allocateAndSetForNonInstrOperand(StructLoweringState &pass,
+                                             SILValue value,
+                                             SILInstruction *applyInst) {
+  assert(value->getType().isObject());
+  SILArgument *arg = dyn_cast<SILArgument>(value);
+  assert(arg && "non-instr operand must be an argmuent");
```

Comment:
We should name the function accordingly. This would make it easier to read at the call-site too.
s/allocateAndSetForNonInstrOperand/allocateAndSetForArgumentOperand/